lib/graphql/schema/timeout_middleware.rb

```diff
@@ -0,0 +1,63 @@
+module GraphQL
+  class Schema
+    # This middleware will stop resolving new fields after `max_seconds` have elapsed.
+    # After the time has passed, any remaining fields will be `nil`, with errors added
+    # to the `errors` key. Any already-resolved fields will be in the `data` key, so
+    # you'll get a partial response.
+    #
+    # Note that this will stop a query _in between_ field resolutions, but
+    # it doesn't interrupt long-running `resolve` functions. Be sure to use
+    # timeout options for external connections. For more info, see
+    # www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/
+    #
+    # @example Stop resolving fields after 2 seconds
+    #   MySchema.middleware << GraphQL::Schema::TimeoutMiddleware.new(max_seconds: 2)
+    #
+    class TimeoutMiddleware
+      # This key is used for storing timeout data in the {Query::Context} instance
+      DEFAULT_CONTEXT_KEY = :__timeout_at__
+      # @param max_seconds [Numeric] how many seconds the query should be allowed to resolve new fields
+      # @param context_key [Symbol] what key should be used to read and write to the query context
+      def initialize(max_seconds:, context_key: DEFAULT_CONTEXT_KEY)
```

Comment:
`max_seconds ` should almost certainly not be kwarg, or at least, it should have a default value if your prefer it like this. On line 27 below, `Time.now + nil` would blow the world up.